### PR TITLE
1.2.x [ELY-1526] Add back the ability for security providers to be loaded using the service loader mechanism but ensure that we skip any provider that is already an installed provider

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.jboss.org/browse/ELY-1526
https://issues.jboss.org/browse/WFLY-9899

### CODE REVIEW

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * A supplier which uses a service loader to find all {@link Provider} instances that aren't in the list of
+ * installed security providers and returns them as an array. The result is then cached.
+ *
+ * @author <a href="mailto:fjuma@redhat.com">Farah Juma</a>
+ * @since 1.2.2
+ */
+public class ProviderServiceLoaderSupplier extends ServiceLoaderSupplier<Provider> {
+
+    private static final HashSet INSTALLED_PROVIDERS_SET;
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: This raw type should have a type of ``HashSet<Class<?>>`` or ``Set<Class<?>>``.

Also it should not be static.  The set of installed providers may change over time, so the set of types of providers can also change over time.  Therefore the set should be reconstructed each time the ``INSTALLED_PROVIDERS`` supplier is evaluated via its ``get()`` method.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        super(Provider.class, classLoader);
+    }
+
+    protected Provider[] loadServices(final Class<Provider> service, final ClassLoader classLoader) {
+        ArrayList<Provider> list = new ArrayList<>();
+        ServiceLoader<Provider> loader = ServiceLoader.load(service, classLoader);
+        Iterator<Provider> iterator = loader.iterator();
+        for (;;) try {
+            if (! iterator.hasNext()) {
+                return list.toArray(Arrays2.createArray(service, list.size()));
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: I don't remember if @darranl has a different recommendation here, but AFAIK accepted best practice is to pass in a constant 0-sized array to ``Collection.<T>toArray(T[] array)`` methods.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            Provider provider = iterator.next();
+            if (! INSTALLED_PROVIDERS_SET.contains(provider.getClass())) {
+                list.add(provider);
+            }
+        } catch (ServiceConfigurationError ignored) {
+            // explicitly ignored
+        }
+    }
+
+    public int hashCode() {
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: The ``hashCode`` impl should match the ``equals`` impl, that is, it should return the hash code of ``classLoader`` or a value derived directly from it.

File: src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java [hunk 4]
```diff
@@ -39,7 +39,7 @@
 public class ServiceLoaderSupplier<E> implements Supplier<E[]> {
 
     private final Class<E> service;
-    private final ClassLoader classLoader;
+    protected final ClassLoader classLoader;
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java]: This is an API change; I think we may want to keep this package-private unless we have a need to open it to external subclasses.

File: src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java [hunk 5]
```diff
@@ -126,9 +128,11 @@
  */
 public final class ElytronXmlParser {
 
-    private static final Supplier<Provider[]> ELYTRON_PROVIDER_SUPPLIER = WildFlySecurityManager.isChecking() ?
-            AccessController.doPrivileged((PrivilegedAction<ServiceLoaderSupplier<Provider>>) () -> new ServiceLoaderSupplier<>(Provider.class, ElytronXmlParser.class.getClassLoader())) :
-            new ServiceLoaderSupplier<>(Provider.class, ElytronXmlParser.class.getClassLoader());
+    private static final Supplier<Provider[]> ELYTRON_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
+            () -> new Provider[] { new WildFlyElytronProvider() },
+            WildFlySecurityManager.isChecking() ?
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java]: I don't think this needs to be privileged since the provider supplier doesn't seem to capture the ACC on construction, unless it does and I just can't see it in this diff...

Comment (fjuma) [reply src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java]: This was switched over to being privileged in ELY-1349: https://github.com/wildfly-security/wildfly-elytron/pull/972

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 3 above)
Comment (fjuma) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: ```super.hashCode()``` does currently return a value derived from the ```classLoader```:

https://github.com/wildfly-security/wildfly-elytron/blob/master/src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java#L82-L89

Does that look ok?

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 2 above)
Comment (fjuma) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: This is actually just a copy/paste from ```ServiceLoaderSupplier#loadServices```. This isn't always 0-sized, this is the result after iterating over the providers.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 1 above)
Comment (fjuma) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: Good catch, will fix this.

**Comment by fjuma** [comment]
Updated based on feedback from David.

File: src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java (hunk 5 above)
Comment (fjuma) [reply src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java]: The parent class, ```ServiceLoaderSupplier```, captures the ACC on construction.

Comment (dmlloyd) [reply src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java]: OK that makes sense.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 2 above)
Comment (dmlloyd) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: ``ServiceLoaderSupplier`` is generic; that's why it uses this array creation strategy.  But since this class is not generic, it does not need to use the dynamic creation strategy.  That said, this isn't incorrect, it's just non-optimal.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 3 above)
Comment (dmlloyd) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: Yeah that looks OK.

File: src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java [hunk 6]
```diff
@@ -65,7 +65,7 @@ public ServiceLoaderSupplier(final Class<E> service, final ClassLoader classLoad
         return result.clone();
     }
 
-    private E[] loadServices(final Class<E> service, final ClassLoader classLoader) {
+    protected E[] loadServices(final Class<E> service, final ClassLoader classLoader) {
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/util/ServiceLoaderSupplier.java]: I think we should not make this ``protected``; it would needlessly increase our API surface area (and there's no guarantee we wouldn't want to refactor at this level).  Package-private is probably the better choice here.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 2 above)
Comment (fjuma) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: Ah, yes, that was silly :) I'll fix that.

**Comment by fjuma** [comment]
Updated based on the latest feedback from @dmlloyd and also made the following two changes to address the two issues @mchoma had mentioned on WFLY-9899:

* Updated ```AuthenticationConfiguration#useProvidersFromClassLoader``` to ensure ```ProviderServiceLoaderSupplier``` is used

* Updated handling of ```use-service-loader" in ```ElytronXmlParser``` to ensure ```ProviderServiceLoaderSupplier``` is used

**Comment by fjuma** [comment]
Updated ```ProviderServiceLoaderSupplier``` to only treat the ```WildFlyElytronProvider``` as part of the list of installed providers if we know it's been added explicitly:

**Comment by hkalina** [comment]
@dmlloyd I should ask you yet, is merging ok, as @fjuma say she has added small second commit yet.

**Comment by dmlloyd** [comment]
Yeah still looks OK.

**Comment by fjuma** [comment]
Thanks @dmlloyd and @hkalina.

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java [hunk 7]
```diff
@@ -39,14 +39,23 @@
  */
 public class ProviderServiceLoaderSupplier extends ServiceLoaderSupplier<Provider> {
 
+    final boolean elytronProviderStaticallyAdded;
```
Comment (mchoma) [inline src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: Why was that necessary to introduce?

File: src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java [hunk 8]
```diff
@@ -168,8 +168,8 @@
     private static final Supplier<Provider[]> DEFAULT_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
             () -> new Provider[] { new WildFlyElytronProvider() },
```
Comment (mchoma) [inline src/main/java/org/wildfly/security/auth/client/AuthenticationConfiguration.java]: With elytronProviderStaticallyAdded=true this is redundant?

File: src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java [hunk 9]
```diff
@@ -130,8 +130,8 @@
     private static final Supplier<Provider[]> ELYTRON_PROVIDER_SUPPLIER = ProviderUtil.aggregate(
             () -> new Provider[] { new WildFlyElytronProvider() },
```
Comment (mchoma) [inline src/main/java/org/wildfly/security/auth/client/ElytronXmlParser.java]: With elytronProviderStaticallyAdded=true this is redundant?

File: src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java (hunk 7 above)
Comment (hkalina) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: For cases when will be Elytron provider added AFTER - like https://github.com/wildfly-security/wildfly-elytron/pull/1101/files#diff-c326ff349cb9d8a1a54a2f94e9e3683aL167

Comment (fjuma) [reply src/main/java/org/wildfly/security/util/ProviderServiceLoaderSupplier.java]: @mchoma This was added because ```ProviderServiceLoaderSupplier``` is now also being used from ```AuthenticationConfiguration#useProvidersFromClassLoader``` and in the handling of ```<use-service-loader module-name=...>```. In these cases, we shouldn't just filter out ```WildFlyElytronProvider``` if it is found since it wasn't already added explicitly by us like it is for the ```DEFAULT_PROVIDER_SUPPLIER```.
